yt/utilities/lib/image_utilities.pyx

```diff
@@ -7,28 +7,132 @@ Utilities for images
     cdef int ys = buffer.shape[1]
+
+    cdef np.float64_t[:, :] buffer_view = buffer
+    cdef np.uint8_t[:, :] buffer_mask_view = buffer_mask
+
     for pi in range(npx):
-        j = <int> (xs * px[pi])
-        i = <int> (ys * py[pi])
-        buffer[i, j] += pv[pi]
-        buffer_mask[i, j] = 1
+        _add_point_to_greyscale_image(buffer_view, buffer_mask_view, px[pi], py[pi], pv[pi], xs, ys)
     return
 
+
+@cython.boundscheck(False)
+def add_points_to_greyscale_image_with_lagrangian_tesselation(
+    np.ndarray[np.float64_t, ndim=2] buffer,
+    np.ndarray[np.uint8_t,   ndim=2] buffer_mask,
+    np.ndarray[np.float64_t, ndim=4] p3d,
+    np.ndarray[np.float64_t, ndim=3] pv,
+    int split=100,
+):
+    cdef int xs = buffer.shape[0]
+    cdef int ys = buffer.shape[1]
+    cdef int Ngrid = p3d.shape[0] - 1
+
+    cdef np.float64_t[:, :] buffer_view = buffer
+    cdef np.uint8_t[:, :] buffer_mask_view = buffer_mask
+
+    cdef np.ndarray[np.uint8_t, ndim=2] vert = np.array(
+        (
+            (0, 0, 0),
+            (1, 0, 0),
+            (1, 1, 0),
+            (0, 1, 0),
+            (0, 0, 1),
+            (1, 0, 1),
+            (1, 1, 1),
+            (0, 1, 1),
+        ),
+        dtype=np.uint8
+    )
+    cdef np.ndarray[np.uint8_t, ndim=2] conn = np.array(
+        (
+            (4, 0, 7, 1),
+            (1, 0, 3, 7),
+            (5, 1, 4, 7),
+            (2, 3, 7, 1),
+            (1, 5, 6, 7),
+            (2, 6, 1, 7),
+        ),
+        dtype=np.uint8
+    )
+
+    cdef np.float64_t[3] orig
+    cdef np.float64_t[3] a
+    cdef np.float64_t[3] b
+    cdef np.float64_t[3] c
+
+    cdef np.float64_t[3] newp
+
+    cdef int i, j, k, m, s, idim
+
+    cdef np.float64_t f = 1.0 / split
+
+    cdef np.uint8_t[:, :] off
+    cdef np.float64_t[:, :] ro
+
+    for m in range(len(conn)):
+        ro = np.random.random(size=(split, 3)) # random offsets
+        off = vert[conn[m]]
+
+        for i in range(Ngrid):
+            for j in range(Ngrid):
+                for k in range(Ngrid):
+                    for idim in range(2):   # stop at 2, because z is not used
+                        orig[idim] = p3d[i + off[3][0], j + off[3][1], k + off[3][2], idim]
+                        a[idim] = p3d[i + off[0][0], j + off[0][1], k + off[0][2], idim] - orig[idim]
+                        b[idim] = p3d[i + off[1][0], j + off[1][1], k + off[1][2], idim] - orig[idim]
+                        c[idim] = p3d[i + off[2][0], j + off[2][1], k + off[2][2], idim] - orig[idim]
+
+                    for s in prange(split, nogil=True):
+                        for idim in range(2):  # stop at 2, because z is not used
+                            newp[idim] = orig[idim] + (
+                                ro[s, 0] * a[idim] +
+                                ro[s, 1] * b[idim] +
+                                ro[s, 2] * c[idim]
+                            )
+                            while newp[idim] < 0.0:
```

Comment:
Can't this be achieved with a single modulus operation ?

Reply:
It can and I've now fixed this!